src/dvc_data/hashfile/cache.py

```diff
@@ -63,3 +65,50 @@ def __init__(
 
     def __getstate__(self):
         return (*super().__getstate__(), self._type)
+
+    def get_many(self, keys):
+        for chunk in chunks(999, keys):
+            select = (
+                "SELECT key, value FROM Cache WHERE key IN (%s) and raw = 1"
+                % ",".join("?" * len(chunk))
+            )
+            rows = self._sql(select, chunk).fetchall()
+
+            chunk = set(chunk)
+            for (key, value) in rows:
+                chunk.remove(key)
+                yield (key, value)
+
+            yield from ((key, None) for key in chunk)
+
+    def set_many(self, d, expire=None, retry=False):
+        if not d:
+            return
+
+        raw = True
+        access_time = store_time = now = time.time()
+        expire_time = None if expire is None else now + expire
+        access_count, tag, size, mode, filename = 0, None, 0, 1, None
+        with self._transact(retry):
+            self._con.executemany(
+                "INSERT OR REPLACE INTO Cache("
+                " key, raw, store_time, expire_time, access_time,"
+                " access_count, tag, size, mode, filename, value"
+                ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
```

Comment:
Btw, why do we use transact in one place but not the other and also use chunks in one place but not the other?

Reply:
And how much slower is normal transact + set/get?

Reply:
The `get_many` is a read, so we don't need a transaction there. We are generating a statement that looks like following:
```sql
SELECT key, value
FROM Cache
WHERE key IN (?, ?, ?, ?, ?, ?, ?, ? ... ?)  # maximum `?` 999 chars
and raw = 1;
```

Those `?` are called parameters, and are limited to 32,766 in the latest versions. For maximum compatibility, I chose 999. I did experiment with numbers >999, but they were always some milliseconds slow, within a margin of error, but reproducible every time.
 
> SQLite allocates space to hold all host parameters between 1 and the largest host parameter number used. Hence, an SQL statement that contains a host parameter like ?1000000000 would require gigabytes of storage. This could easily overwhelm the resources of the host machine. To prevent excessive memory allocations, the maximum value of a host parameter number is SQLITE_MAX_VARIABLE_NUMBER, which defaults to 999 for SQLite versions prior to 3.32.0 (2020-05-22) or 32766 for SQLite versions after 3.32.0.

`executemany` caches the statement too and would have simplified this because we could pass without looping, but it does not support `SELECT` statements and does not return values.

Regarding `set_many`, we do need a transaction because it's a write. `executemany` supports iterators, so we are taking advantage of that here instead of chunking. I did experiment with chunking too but did not see any difference.

Reply:
Here's a raw benchmark results:

| items   | get_many   | get_many_with_get   | set_many   | set_many_with_set   | set_many_with_set_transact   |
|---------|------------|---------------------|------------|---------------------|------------------------------|
| 10      | 19.98 mks  | 64.95 mks           | 63.46 mks  | 369.89 mks          | 243.69 mks                   |
| 100     | 126.54 mks | 711.12 mks          | 537.82 mks | 3.66 ms             | 2.29 ms                      |
| 1,000    | 1.42 ms    | 6.65 ms             | 6.01 ms    | 41.32 ms            | 25.33 ms                     |
| 10,000   | 16.64 ms   | 70.31 ms            | 69.53 ms   | 411.75 ms           | 249.74 ms                    |
| 100,000  | 181.54 ms  | 717.86 ms           | 704.89 ms  | 4.11 s              | 2.53 s                       |
| 1,000,000 | 2.67 s     | 8.55 s              | 10.10 s    | 43.52 s             | 29.53 s                      |